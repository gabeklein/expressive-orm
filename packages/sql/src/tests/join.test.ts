import { Num, Query, Str, Table, Type } from '../';

class Foo extends Type {
  name = Str();
  color = Str();
}

class Bar extends Type {
  name = Str();
  color = Str();
  rating = Num();
}

class Baz extends Type {
  color = Str();
  rating = Num();
}

it("will join using object", async () => {
  const query = Query(where => {
    const foo = where(Foo);
    const bar = where(Bar, { color: foo.color });
    const baz = where(Baz, { rating: bar.rating });

    where(foo.name).isNot("Danny");
    where(bar.rating).isMore(50);
    where(baz.color).is("blue");
  });

  expect(query).toMatchSnapshot();
})

it("will join using function", async () => {
  const query = Query(where => {
    const foo = where(Foo);
    const bar = where(Bar, on => {
      on(bar.name).isNot(foo.name);
      on(bar.color).is(foo.color);
    });

    where(foo.name).isNot("Danny");
    where(bar.rating).isMore(50);
  });

  expect(query).toMatchSnapshot();
})

it("will alias tables which have a schema", () => {
  class Foo extends Type {
    this = Table({ schema: "foobar" });

    name = Str();
    color = Str();
  }

  const query = Query(where => {
    const foo = where(Foo);

    where(foo.color).is("red");
  })

  expect(query).toMatchSnapshot();
})