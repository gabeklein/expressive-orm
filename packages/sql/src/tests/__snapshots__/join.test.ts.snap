// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will alias tables which have a schema 1`] = `
	select
	  count(*)
	from
	  \`foobar\`.\`foo\` as \`$0\`
	where
	  \`$0\`.\`color\` = 'red'
`;

exports[`will join using function 1`] = `
	select
	  count(*)
	from
	  \`foo\`
	  inner join \`bar\` on \`bar\`.\`name\` <> \`foo\`.\`name\`
	  and \`bar\`.\`color\` = \`foo\`.\`color\`
	where
	  \`foo\`.\`name\` <> 'Danny'
	  and \`bar\`.\`rating\` > 50
`;

exports[`will join using object 1`] = `
	select
	  count(*)
	from
	  \`foo\`
	  inner join \`bar\` on \`bar\`.\`color\` = \`foo\`.\`color\`
	  inner join \`baz\` on \`baz\`.\`rating\` = \`bar\`.\`rating\`
	where
	  \`foo\`.\`name\` <> 'Danny'
	  and \`bar\`.\`rating\` > 50
	  and \`baz\`.\`color\` = 'blue'
`;
